elasticdl/python/ps/parameters.py

```diff
@@ -175,10 +175,13 @@ def to_model_pb(self):
         """
         model_pb = elasticdl_pb2.Model()
         model_pb.version = self.version
-        for name, var in self.non_embedding_params.items():
+        non_embedding_params = self.non_embedding_params.copy()
+        for name, var in non_embedding_params.items():
             serialize_ndarray(var.numpy(), model_pb.dense_parameters[name])
+        del non_embedding_params
 
-        for name, embedding_table in self.embedding_params.items():
+        embedding_params = self.embedding_params.copy()
+        for name, embedding_table in embedding_params.items():
```

Comment:
Maybe this embedding_param iteration is not the cause of the bug, as `self.embedding_params` has a fixed number (number of embedding tables) after initialization.

It is probably in `embedding_table.to_indexed_slices(),` a few lines later, as when iterating through the embedding vectos, new vectors may be added by other threads.

Reply:
Yes. It is more suitable to copy `embedding_table.embedding_vectors` before iteration.

Reply:
embedding_table.embedding_vectors may be too long for copy. If lock does not affect training efficiency when checkpoint is not saved, better to use a lock.

Reply:
In PS with Python, the `optimizer_wrapper` has used a lock to update the embedding table. https://github.com/sql-machine-learning/elasticdl/blob/e695ba7b33f01383798b28825fab408dae4e1327/elasticdl/python/ps/optimizer_wrapper.py#L155-L157

So, it does not affect training efficiency to use a lock for `set` and `get` in `EmbeddingTable`. I will use a lock in the next commit.